test/server/http_filter_integration_test_base.h

```diff
@@ -0,0 +1,122 @@
+   * "{{response_body_size:1024}".
+   * @return Envoy::IntegrationStreamDecoderPtr Pointer to the integration stream decoder, which can
+   * be used to inspect the response.
+   */
+  Envoy::IntegrationStreamDecoderPtr
+  getResponseFromUpstream(absl::string_view request_level_config);
+
+  /**
+   * @brief Fetches a response with request-level configuration set in the request header. The
+   * extension under test should supply the response.
+   *
+   * @param request_level_config Configuration to be delivered by request header. For example
+   * "{{response_body_size:1024}".
+   * @return Envoy::IntegrationStreamDecoderPtr Pointer to the integration stream decoder, which can
+   * be used to inspect the response.
+   */
+  Envoy::IntegrationStreamDecoderPtr
+  getResponseFromExtension(absl::string_view request_level_config);
+
+  /**
+   * @brief Fetch a response using the provided request headers, and set up a fake upstream to
+   * supply a response.
+   *
+   * @param request_headers Supply a full set of request headers. If the request method is set to
+   * POST, an entity body will be send after the request headers.
+   * @return Envoy::IntegrationStreamDecoderPtr Pointer to the integration stream decoder, which can
+   * be used to inspect the response.
+   */
+  Envoy::IntegrationStreamDecoderPtr
+  getResponseFromUpstream(const Envoy::Http::TestRequestHeaderMapImpl& request_headers);
+
+  /**
+   * @brief Fetch a response using the provided request headers. The extension under test must
+   * supply a response.
+   *
+   * @param request_headers Supply a full set of request headers. If the request method is set to
+   * POST, an entity body will be send after the request headers.
+   * @return Envoy::IntegrationStreamDecoderPtr Pointer to the integration stream decoder, which can
+   * be used to inspect the response.
+   */
+  Envoy::IntegrationStreamDecoderPtr
+  getResponseFromExtension(const Envoy::Http::TestRequestHeaderMapImpl& request_headers);
+
+private:
+  /**
+   * @brief Fetches a response with request-level configuration set in the request header.
+   *
+   * @param request_level_config Configuration to be delivered by request header. For example
+   * "{{response_body_size:1024}".
+   * @param setup_for_upstream_request Set to true iff the filter extension under test is expected
+   * to short-circuit and supply a response directly. For example because it couldn't parse the
+   * supplied request-level configuration. Otherwise this should be set to true, and a stock
+   * response will be yielded by the integration test server through an upstream request.
+   * @return Envoy::IntegrationStreamDecoderPtr Pointer to the integration stream decoder, which can
+   * be used to inspect the response.
+   */
+  Envoy::IntegrationStreamDecoderPtr getResponse(absl::string_view request_level_config,
+                                                 bool setup_for_upstream_request);
```

Comment:
Could we reduce the number of public methods by half and eliminate the need to figure out the meaning of `true` or `false` in the implementation by creating a mode enum with values `kUpstream` and `kExtension`? Or if it's more convenient for the test authors to have the separate public methods, a private enum could still make the impl easier to read.